documentation/dataplane/dash-flow-resimulation.md

```diff
@@ -0,0 +1,213 @@
+- Each flow also stores the current id when it is created.
+
+### 1.2. Flow resimulation process
+
+With flow incarnation id, we can implement full flow resimulation as follows:
+
+- When a new packet arrives, the current flow incarnation id will be stored into the metadata bus.
+- In Conntrack Lookup stage, if we found the id stored in its flow doesn't match the one in the pipeline, we treat this flow as resimulated.
+- For simulated flows, we continue to run the later ACL and match stages instead of directly applying the actions saved in the flow.
+- In Conntrack Update stage, a new pair of flow will be generated and replaces the current flow, which finishes the process of flow resimulation.
+
+After this, the later packet will directly hit the new flow and bypass the later stages.
+
+### 1.3. Flow incarnation id overflow
+
+Some flows can be very low volume and causing flow incarnation id stops working. Say, if the flow incarnation id is 8 bits, and not it is set to 1, a low volume flow is created. And after 256 flow resimulation calls, the next packet finally arrives and see the same id, which bypasses the flow resimulation process.
+
+To solve this, whenever the incarnation id overflows, we will need to treat all flows as resimulated. Implementation-wise, this can be done by adding another bit to indicate overflow happened, then reset the flow incarnation id to 0 in each flow during the next flow aging process. After flow aging is done, we can reset the overflow bit.
+
+Since frequently calling full flow resimulation is not going to be a good practice, this will be a rare case, so we don't need to worry about the performance impact of this.
+
+### 1.4. Flow resimulation w/ flow HA
+
+When flow HA is enabled, both flow incarnation id and the updated flow needs to be synced to the standby pipeline.
+
+The process follows the flow HA design. When replacing the current flow with the newly generated flows, the new flows will be marked as "not-synced". Then the packet will be sent back to the pipeline, which triggers the inline flow sync logic and takes the flow information to the other side.
+
+Since the flow incarnation id is part of the flow state, it will also be synched to the standby side and updates the stored id there. Although in flow HA design, the policy will be programmed to both active and standby side, but we cannot use the id in the standby side directly, because these 2 pipelines are programmed independently, so we don't have a way to ensure that the standby one always matches the active one. To solve this, we make the standby pipeline always follows the active side, which follows the flow lifetime management design in flow HA.
+
+### 1.5. Object model change summary for full flow resimulation
+
+To summarize, the following changes are needed to implement full flow resimulation:
+
+- 2 properties needs to be added for each pipeline:
+
+    ```json
+    "DASH_SAI_ENI_TABLE|123456789012": {
```

Comment:
Section 1.1 talks about pipeline level ID, while the schema defines per-ENI ID. Which one's correct? I would prefer a more specific per-ENI if it's possible, to reduce the blast radius of the ID increment.

Reply:
updated.